### PR TITLE
Metadata format deploys

### PR DESCRIPTION
**Pull request by mdonnalley**
### What does this PR do?

Add `--metadata-dir` to `deploy metadata` and `deploy metadata validate`

Depends on https://github.com/forcedotcom/source-deploy-retrieve/pull/619

TODO:
- [x] NUTs
- [ ] Review UX
- [ ] Review messages

### What issues does this PR fix or reference?
@W-10748958@

### CODE REVIEW

File: src/utils/flags.ts [hunk 1]
```diff
@@ -0,0 +1,43 @@
+/*
```
Comment (mshanemc) [inline src/utils/flags.ts:1]: I like this pattern a lot

File: src/utils/flags.ts [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+): Interfaces.OptionFlag<TestLevel | undefined> => {
+  return Flags.build<TestLevel | undefined>({
+    char: 'l',
+    parse: (input: string) => Promise.resolve(input as TestLevel),
+    options: Object.values(TestLevel),
+    ...opts,
+  })();
+};
+
+const parsePathInfo = async (input: string, exists: boolean): Promise<PathInfo> => {
```
Comment (mshanemc) [inline src/utils/flags.ts:22]: suggestion: `mustExist` would be clearer

**Review by mshanemc — COMMENTED** [review]
1. good use of custom flag (fileOrDirectoryFlag)
2. this should be shared/demo'd as an example of making a "richer" flag that has extra properties about it that you can refer to later

**Review by mshanemc — APPROVED** [review]

**Comment by mshanemc** [comment]
QA Notes:

### setup
`git clone https://github.com/trailheadapps/dreamhouse-lwc dh-md-deploy`
`sfdx force:source:convert --outputdir=mdapiOut --sourcepath=force-app` produces a decent package.xml, folder

`../../plugin-deploy-retrieve/bin/dev deploy metadata --metadata-dir mdapiOut`
❓ this just hangs for a really long time (is pending in the org, though, so it did start and get a jobId).  Maybe that's a server problem?
Screen Shot 2022-05-10 at 10 18 15 AM

_laptop went to sleep waiting for it.  Good chance to try a resume_
`resume` doesn't seem to get any components and its progress bar also seems stuck.

❌  using `../../plugin-deploy-retrieve/bin/dev deploy metadata resume --use-most-recent` is creating a new deployment 

(see ids and screenshot from the org) instead of resuming the in progress one 
Screen Shot 2022-05-10 at 10 32 00 AM
Screen Shot 2022-05-10 at 10 32 09 AM

maybe I can report using an ID
`../../plugin-deploy-retrieve/bin/dev deploy metadata report --job-id 0AfJ0000029EvSg` (id was copy-pasted from the org)
❌ Error (InvalidJobIdError): No job found for ID: 0AfJ0000029EvSg.

✅ `../../plugin-deploy-retrieve/bin/dev deploy metadata report --job-id 0AfJ0000029EvSgKAK` (uses 18char id instead of 15) and `--json` of same yield good output

**Comment by mdonnalley** [comment]
@mshanemc 

> maybe I can report using an ID
../../plugin-deploy-retrieve/bin/dev deploy metadata report --job-id 0AfJ0000029EvSg (id was copy-pasted from the org)
❌ Error (InvalidJobIdError): No job found for ID: 0AfJ0000029EvSg.

We've previously agreed upon not supporting the resuming of deployments that are not in the deploy cache. Is that something that you're reconsidering now?

**Comment by mshanemc** [comment]
> @mshanemc
> 
> > maybe I can report using an ID
> > ../../plugin-deploy-retrieve/bin/dev deploy metadata report --job-id 0AfJ0000029EvSg (id was copy-pasted from the org)
> > ❌ Error (InvalidJobIdError): No job found for ID: 0AfJ0000029EvSg.
> 
> We've previously agreed upon not supporting the resuming of deployments that are not in the deploy cache. Is that something that you're reconsidering now?

Maybe I'm not following--this comment was about `report` (not resume) only working with 18 char id, not 15.

**Review by mshanemc — APPROVED** [review]

**Comment by mshanemc** [comment]
✅  resume is fixed (tested with --async on both zip file and dir)

usability: I'm rethinking this now that I'm using it
```sh
➜  dh-md-deploy git:(main) ✗ ../../plugin-deploy-retrieve/bin/dev deploy metadata report --job-id 0AfJ0000029b3CpKAI
Warning: This command is currently in beta. Any aspect of this command can change without advanced notice. Don't use beta commands in your scripts.
Error (InvalidJobIdError): No job found for ID: 0AfJ0000029b3CpKAI.
```

That's not gonna be obvious to users without an understanding of what the cache is and how it works--they'll be like, "it's totally a valid Id, I see it there in the org"

One option would be to not remove them from the cache until they expire
Another would be to mark them as completed in the cache so that you could still find them *if you ask for them by ID* but they wouldn't be used by `--use-most-recent`

thoughts?

**Comment by peternhale** [comment]
> ✅ resume is fixed (tested with --async on both zip file and dir)
> 
> usability: I'm rethinking this now that I'm using it
> 
> ```shell
> ➜  dh-md-deploy git:(main) ✗ ../../plugin-deploy-retrieve/bin/dev deploy metadata report --job-id 0AfJ0000029b3CpKAI
> Warning: This command is currently in beta. Any aspect of this command can change without advanced notice. Don't use beta commands in your scripts.
> Error (InvalidJobIdError): No job found for ID: 0AfJ0000029b3CpKAI.
> ```
> 
> That's not gonna be obvious to users without an understanding of what the cache is and how it works--they'll be like, "it's totally a valid Id, I see it there in the org"
> 
> One option would be to not remove them from the cache until they expire Another would be to mark them as completed in the cache so that you could still find them _if you ask for them by ID_ but they wouldn't be used by `--use-most-recent`
> 
> thoughts?

I don't think changing cache behavior to resolve this issue is complete enough. Retaining cache entry until expiration is just delaying exposure of the same issue. If using a config to track deploy state past deploy complete is desired, then it should be treated as a permanent config and not use TTL.

These are the possibilities if job-id is supplied and a cache is not found:
- job-id does not exist
- Wrong job-id was provided (meaning the deploy was started by someone else)
- job-id is correct, job is not complete, but the cache expired
- job-id is correct, job is complete and cache entry was removed

If cache entry is missing, can a resume be successful using data fetched from org using job-id? If so, why not support this possibility?

At a minimum, I think that when the resume encounters a cache miss when user provides a job-id it should verify the job-id against the org so a proper status can be communicated to the user; job-id not found, deploy status and if resume is not possible using data from the org, inform the user of that too.

**Pull request merged**